keps/sig-cli/3515-kubectl-explain-openapiv3/README.md

```diff
@@ -0,0 +1,909 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes an enhancement to kubectl explain:
+
+1. Switch data source from OpenAPI v2 to OpenAPI v3
+2. Replace the hand-written `kubectl explain` printer with a go/template implementation.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### OpenAPI v3 is a richer API description than OpenAPI v2
+
+OpenAPI v3 support in Kubernetes is currently beta since version 1.24.
+OpenAPI V3 is a richer representation of the kubernetes API to our users, who have been asking for visibility
+into things like:
+
+1. nullable
+2. default
+3. validation fields like oneOf, anyOf, etc.
+
+To show each of these additional data points by themselves is a strong reason
+to switch to using OpenAPI v3.
+
+
+### CRD schemas expressed as OpenAPI v2 are lossy
+
+Today CRDs specify their schemas in OpenAPI v3 format. To serve the `/openapi/v2`
+document used today by kubectl, there is an expensive conversion from the v3 down
+to v2 format.
+
+This process is [very lossy](https://github.com/kubernetes/kubernetes/blob/6e0de20fbb4c127d2e45c7a22347c08545fc7a86/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/v2/conversion.go#L56-L66), so `kubectl explain` when used against CRDs
+making use of v3 features does not have a good experience with inaccurate information, or fields removed altogther.
+
+This transformation causes bugs, for example, when attempting to `explain` a field
+that is `nullable`, kubectl instead shows nothing, due to the lossy conversion
+wiping nullable fields.
+
+### Goals
+
+1. Provide the new richer type information specified by OpenAPI v3 within kubectl explain
+2. Have a more maintainable `text/template` based approach to printing
+3. Fallback to old `explain` implementation if cluster does not expose OpenAPI v3 data.
+4. Provide multiple new output formats for kubectl explain:
+    * human-readable plaintext
+    * markdown
+    * maybe others
+5. (Optional?) Allow users to specify their own templates for use with kubectl
+  explain (there may be interesting use cases for this)
```

Comment:
Not saying this should be in scope for this KEP, but I think there's an interesting tie-in here to the ideas in https://github.com/kubernetes/kubectl/issues/914. That issue proposes embedding templates that can be used to generate sample resources directly in the openAPI, so that instead of bespoke (and therefore unmaintainable) `kubectl create` commands, we could have fully extensible `kubectl generate` commands driven by this data. If `kubectl explain` is used to drive documentation in an even richer way thanks to this KEP, it would be pretty awesome if it could include generated examples in its output, perhaps through that mechanism.